packages/shopify-app-session-storage-dynamodb/src/dynamodb.ts

```diff
@@ -0,0 +1,131 @@
+import {marshall, unmarshall} from '@aws-sdk/util-dynamodb';
+import {Session, SessionParams} from '@shopify/shopify-api';
+import {SessionStorage} from '@shopify/shopify-app-session-storage';
+
+export interface DynamoDBSessionStorageOptions {
+  sessionTableName: string;
+  shopIndexName: string;
+  config?: DynamoDBClientConfig;
+}
+
+const defaultDynamoDBSessionStorageOptions: DynamoDBSessionStorageOptions = {
+  sessionTableName: 'shopify_sessions',
+  shopIndexName: 'shop_index',
+};
+
+export class DynamoDBSessionStorage implements SessionStorage {
+  private client: DynamoDBClient;
+  private options: DynamoDBSessionStorageOptions;
+
+  constructor(opts?: DynamoDBSessionStorageOptions) {
+    this.options = {...defaultDynamoDBSessionStorageOptions, ...opts};
+    this.client = new DynamoDBClient({...this.options.config});
+  }
+
+  public async storeSession(session: Session): Promise<boolean> {
+    await this.client.send(
+      new PutItemCommand({
+        TableName: this.options.sessionTableName,
+        Item: this.serializeSession(session),
+      }),
+    );
+
+    return true;
+  }
+
+  public async loadSession(id: string): Promise<Session | undefined> {
+    if (!id) return undefined;
+
+    const result = await this.client.send(
+      new GetItemCommand({
+        TableName: this.options.sessionTableName,
+        Key: this.seriliazeId(id),
+      }),
+    );
+
+    return result.Item ? this.deserilaizeSession(result.Item) : undefined;
+  }
+
+  public async deleteSession(id: string): Promise<boolean> {
+    await this.client.send(
+      new DeleteItemCommand({
+        TableName: this.options.sessionTableName,
+        Key: this.seriliazeId(id),
+      }),
+    );
+
+    return true;
+  }
+
+  public async deleteSessions(ids: string[]): Promise<boolean> {
+    await Promise.all(ids.map((id) => this.deleteSession(id)));
+    return true;
+  }
+
+  public async findSessionsByShop(shop: string): Promise<Session[]> {
+    const result = await this.client.send(
+      new QueryCommand({
+        TableName: this.options.sessionTableName,
+        IndexName: this.options.shopIndexName,
+        KeyConditionExpression: 'shop = :shop',
+        ExpressionAttributeValues: marshall({
+          ':shop': shop,
+        }),
+        ProjectionExpression: 'id, shop',
+      }),
+    );
+
+    const sessions = await Promise.all(
+      result.Items?.map((item) => this.loadSession(this.deserializeId(item))) ||
+        [],
+    );
+
+    return sessions.filter(
+      (session): session is Session => session !== undefined,
+    );
+  }
+
+  private seriliazeId(id: string): {[key: string]: AttributeValue} {
+    return marshall({id});
+  }
+
+  private deserializeId(id: {[key: string]: AttributeValue}): string {
+    return unmarshall(id).id;
+  }
+
+  private serializeSession(session: Session): {[key: string]: AttributeValue} {
+    // DynamoDB doesn't support Date objects, so we need to convert it to a number
```

Comment:
I would rather use a string to store the date, see also https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.NamingRulesDataTypes.html#HowItWorks.DataTypes.String

Reply:
I originally stored it as number to avoid any timezone or DST issues. However, `date.toISOString()` returns the date always in UTC (according to MDN), so that's okay.